server/src/main/java/io/crate/statistics/ReservoirSampler.java

```diff
@@ -183,159 +159,211 @@ Samples getSamples(RelationName relationName, List<Reference> columns, int maxSa
+        LuceneCollectorExpression<?> collector;
+        Input<?> input;
+        final ColumnStatsBuilder<T> statsBuilder;
 
-            try {
-                long bytesRead = sizeEstimator.estimateSize(row);
-                ramAccounting.addBytes(bytesRead);
-                bytesSinceLastPause = maybePause(bytesRead, bytesSinceLastPause);
-            } catch (CircuitBreakingException e) {
-                LOGGER.info(
-                    "Stopped gathering samples for `ANALYZE` operation because circuit breaker triggered. "
-                        + "Generating statistics with {} instead of {} records",
-                    records.size(),
-                    maxSamples
-                );
-                break;
+        private ColumnCollector(ColumnStatsBuilder<T> statsBuilder) {
+            this.statsBuilder = statsBuilder;
+        }
+
+        void setShard(LuceneCollectorExpression<?> collector, Input<?> input) {
+            this.collector = collector;
+            this.input = input;
+        }
+
+        @SuppressWarnings("unchecked")
+        void collect(int docId, RamAccounting ramAccounting) {
+            collector.setNextDocId(docId);
+            statsBuilder.add((T) input.value(), ramAccounting);
+        }
+    }
+
+    private record ExpressionsAndInputs(
+        List<? extends LuceneCollectorExpression<?>> expressions,
+        List<Input<?>> inputs
+    ) {}
+
+    private record ShardExpressions(Engine.Searcher searcher,
+                                    ExpressionsAndInputs expressionsAndInputs) {
+
+        void updateColumnCollectors(List<ColumnCollector<?>> collectors) {
+            assert collectors.size() == expressionsAndInputs.expressions.size();
+            assert collectors.size() == expressionsAndInputs.inputs.size();
+            for (int i = 0; i < collectors.size(); i++) {
+                collectors.get(i).setShard(expressionsAndInputs.expressions.get(i), expressionsAndInputs.inputs.get(i));
             }
-            records.add(new RowN(row));
         }
 
-        return records;
     }
 
-    static class DocIdToRow implements IntFunction<Object[]> {
+    private static class ColumnSampler extends FetchIdIterator {
+
+        final List<ColumnCollector<?>> expressions;
+        final IntFunction<ShardExpressions> shardSupplier;
+        final RamAccounting ramAccounting;
 
-        private final Engine.Searcher searcher;
-        private final List<Input<?>> inputs;
-        private final List<? extends LuceneCollectorExpression<?>> expressions;
+        ShardExpressions currentShard;
+        ReaderContext currentLeafContext;
+        int rowsCollected;
 
-        DocIdToRow(Engine.Searcher searcher,
-                   List<Input<?>> inputs,
-                   List<? extends LuceneCollectorExpression<?>> expressions) {
-            this.searcher = searcher;
-            this.inputs = inputs;
+        private ColumnSampler(
+            List<ColumnCollector<?>> expressions,
```

Comment:
```suggestion
            List<ColumnCollector<?>> collectors,
```

It's a bit confusing to use `expressions` here, which we usually use for instances of `CollectExpression`. And there's also `ShardExpressions` here, together with the use of `var` - makes it difficult to keep track of what's what if everything is expressions.